framework/aster-frame/src/mm/page/mod.rs

```diff
@@ -0,0 +1,197 @@
+    /// `Page`. Otherwise double-free will happen.
+    ///
+    /// Also, the caller ensures that the usage of the page is correct. There's
+    /// no checking of the usage in this function.
+    pub(in crate::mm) unsafe fn restore(paddr: Paddr) -> Self {
+        let vaddr = mapping::page_to_meta::<PagingConsts>(paddr);
+        let ptr = vaddr as *const MetaSlot;
+
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Clone a `Page` handle from a forgotten `Page` as a physical address.
+    ///
+    /// This is similar to [`Page::restore`], but it also increments the reference count
+    /// and the forgotten page will be still leaked unless restored later.
+    ///
+    /// # Safety
+    ///
+    /// The safety requirements are the same as [`Page::restore`].
+    pub(in crate::mm) unsafe fn clone_restore(paddr: &Paddr) -> Self {
+        let vaddr = mapping::page_to_meta::<PagingConsts>(*paddr);
+        let ptr = vaddr as *const MetaSlot;
+
+        let refcnt = unsafe { &(*ptr).refcnt };
+        refcnt.fetch_add(1, Ordering::Relaxed);
+
+        Self {
+            ptr,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Get the physical address.
+    pub fn paddr(&self) -> Paddr {
+        mapping::meta_to_page::<PagingConsts>(self.ptr as Vaddr)
+    }
+
+    /// Get the reference count of this page.
+    ///
+    /// # Safety
+    ///
+    /// This method by itself is safe, but using it correctly requires extra care.
+    /// Another thread can change the reference count at any time, including
+    /// potentially between calling this method and the action depending on the
+    /// result.
+    fn ref_count(&self) -> u32 {
+        self.refcnt().load(Ordering::Relaxed)
+    }
+
+    /// Get the metadata of this page.
+    pub fn meta(&self) -> &M {
+        unsafe { &*(self.ptr as *const M) }
+    }
+
+    /// Get the mutable metadata of this page.
+    ///
+    /// # Safety
+    ///
+    /// The caller should be sure that the page is exclusively owned.
+    pub(in crate::mm) unsafe fn meta_mut(&mut self) -> &mut M {
+        unsafe { &mut *(self.ptr as *mut M) }
+    }
+
+    fn refcnt(&self) -> &AtomicU32 {
+        unsafe { &(*self.ptr).refcnt }
+    }
+}
+
+impl<M: PageMeta> Clone for Page<M> {
+    fn clone(&self) -> Self {
+        self.refcnt().fetch_add(1, Ordering::Relaxed);
+        Self {
+            ptr: self.ptr,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<M: PageMeta> Drop for Page<M> {
+    fn drop(&mut self) {
+        if self.refcnt().fetch_sub(1, Ordering::Release) == 1 {
+            // A fence is needed here with the same reasons stated in the implementation of
+            // `Arc::drop`: <https://doc.rust-lang.org/std/sync/struct.Arc.html#method.drop>.
+            core::sync::atomic::fence(Ordering::Acquire);
+            // Let the custom dropper handle the drop.
+            M::drop(self);
```

Comment:
Calling `M::drop` before `usage.store(0, ...)` is problematic. 

For example, let's consider the case when `M = FrameMeta`. When the internal reference count of a `Frame` drops to 0, this `drop` method is called, which then calls `FrameMeta::drop`. 

```rust
impl PageMeta for FrameMeta {
    const USAGE: PageUsage = PageUsage::Frame;

    fn drop(page: &mut Page<Self>) {
        unsafe { allocator::dealloc(page.paddr() / PAGE_SIZE, 1) };
    }
}
```

The `FrameMeta::drop` function then calls `allocator::dealloc`. After this point of time, a concurrent thread is able to allocate the same frame by calling `allocator::alloc_xxx`. But the success of this step is insufficient for `VmAllocOptions::alloc` as it relies on `Page::from_unused` to return the required instances of `Frame`. But unfortunately, at this point of time `usage.store(0, ...)` is not executed yet, so `Page::from_unused` won't succeed.